Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,470 @@
+evenly into the machine word size. For such types, you have to use
+'newAlignedPinnedPrimArray' to ensure that alignment is respected.
+
+Here is an example of an RGB pixel 'Prim' instance.
+
+@
+  data Pixel = Pixel Word8 Word8 Word8 -- you may want to unpack these
+  instance Prim Pixel where
+    sizeOf# _ = 3#
+    alignment# _ = 1#
+    indexByteArray# ba# i# = Pixel
+      (indexByteArray# ba# i#)
+      (indexByteArray# ba# (i# +# 1#))
+      (indexByteArray# ba# (i# +# 2#))
+    ...
+@
+
+Now you can use 'PrimArray Pixel' with either this module or
+"Data.Primitive.Array.Class/Data.Primitive.Array.Checked".
+-}
+
+
+module Data.Primitive.PrimArray (
+  -- * Types
+  PrimArray(..), MutablePrimArray(..),
+
+  -- * Allocation
+  newPrimArray, newPinnedPrimArray, newAlignedPinnedPrimArray,
+
+  -- * Element access
+  readPrimArray, writePrimArray, indexPrimArray,
+
+  -- * Freezing and thawing
+  unsafeFreezePrimArray, unsafeThawPrimArray,
+
+  -- * Block operations
+  copyPrimArray, copyMutablePrimArray, movePrimArray,
+  setPrimArray, resizeMutablePrimArray, shrinkMutablePrimArray,
+  copyPrimArrayToPtr, copyMutablePrimArrayToPtr, copyMutablePrimArrayFromPtr,
+
+  -- * Information
+  sizeofPrimArray, sizeofMutablePrimArray, sameMutablePrimArray,
+  primArrayContents, mutablePrimArrayContents, withPrimArrayContents, withMutablePrimArrayContents,
+  isPrimArrayPinned, isMutablePrimArrayPinned,
+  samePrimArray,
+
+  -- * Prefetch
+  -- | Check <http://hackage.haskell.org/package/ghc-prim-0.3.1.0/docs/GHC-Prim.html#Prefetch> for more information.
+
+  prefetchPrimArray0, prefetchPrimArray1, prefetchPrimArray2, prefetchPrimArray3,
+  prefetchMutablePrimArray0, prefetchMutablePrimArray1, prefetchMutablePrimArray2, prefetchMutablePrimArray3
+
+) where
+
+
+import Control.Monad.Primitive
+import Control.Monad (liftM)
+import Data.Primitive
+import Data.Typeable
+import Data.Data
+import GHC.Prim
+import Data.Primitive.Internal.Compat ( isTrue# )
+import GHC.Base (Int(..), Char(..))
+import GHC.Float (Float(..), Double(..))
+import GHC.Word (Word(..), Word8(..), Word16(..), Word32(..), Word64(..))
+import GHC.Int (Int8(..), Int16(..), Int32(..), Int64(..))
+import GHC.Ptr (Ptr(..), FunPtr(..))
+
+#if !(MIN_VERSION_base(4,8,0))
+import Data.Monoid (mappend)
+#endif
+
+-- | Primitive array tagged with element type @a@.
+--
+newtype PrimArray a = PrimArray ByteArray
+    deriving (Typeable, Data)
+
+instance (Prim a, Eq a) => Eq (PrimArray a) where
+    {-# INLINE (==) #-}
+    (==) = eqPrimArray
+
+eqPrimArray :: (Prim a, Eq a) => PrimArray a -> PrimArray a -> Bool
+{-# INLINE [1] eqPrimArray #-}
+eqPrimArray paA paB
+    | paA `samePrimArray` paB = True
+    | otherwise =
+        let !sizA = sizeofPrimArray paA
+            !sizB = sizeofPrimArray paB
+        in sizA == sizB && go 0 sizA
+  where
+    go !i !siz | i >= siz  = True
+               | otherwise = indexPrimArray paA i == indexPrimArray paB i && go (i+1) siz
+
```

Comment:
I have read the core before, but have got a test yet. Built-in primitive types should not be handled by this code, there're rules to rewrite the loop into `memcmp`.